src/java/voldemort/utils/pool/KeyedResourcePool.java

```diff
@@ -123,35 +126,53 @@ public V checkout(K key) throws Exception {
         checkNotClosed();
 
         long startNs = System.nanoTime();
+        long timeoutNs = resourcePoolConfig.getTimeout(TimeUnit.NANOSECONDS);
+        long endNs = startNs + timeoutNs;
         Pool<V> resourcePool = getResourcePoolForKey(key);
         V resource = null;
         try {
             checkNotClosed();
-            // Must attempt a non blocking checkout before blockingGet to ensure
-            // resources are created for the pool.
-            resource = attemptNonBlockingCheckout(key, resourcePool);
+            long totalNonBlockingElapsedNs = 0;
+            long iterStartTime = 0;
+            long totalBlockingElapsedNs = 0;
+            final long MAX_WAIT_TIME = 300 * Time.NS_PER_MS;
 
-            if(resource == null) {
-                long nonBlockingElapsedNs = System.nanoTime() - startNs;
-                long timeRemainingNs = resourcePoolConfig.getTimeout(TimeUnit.NANOSECONDS)
-                                       - nonBlockingElapsedNs;
+            while(resource == null && (iterStartTime = System.nanoTime()) < endNs) {
```

Comment:
This is not a spin lock definitely. I try to wait 2/3 times instead of one big wait. Now the connection creation is asynchronous so if one thread triggers the async and waits, other thread could have snatched before the first thread could call wait. So instead of one big wait, now the thread tries to break the time into 2 or 3 chunks and try to do that. Spin lock is not an option and it will burn the CPU like hell. git is not letting me expand the code here, I believe the number is 2 and not 3, but not sure.